enterprise/backend/src/metabase_enterprise/serialization/api.clj

```diff
@@ -0,0 +1,242 @@
+(ns metabase-enterprise.serialization.api
+  (:require
+   [clojure.java.io :as io]
+   [compojure.core :refer [POST]]
+   [java-time.api :as t]
+   [metabase-enterprise.serialization.v2.extract :as extract]
+   [metabase-enterprise.serialization.v2.ingest :as v2.ingest]
+   [metabase-enterprise.serialization.v2.load :as v2.load]
+   [metabase-enterprise.serialization.v2.storage :as storage]
+   [metabase.api.common :as api]
+   [metabase.api.routes.common :refer [+auth]]
+   [metabase.logger :as logger]
+   [metabase.models.serialization :as serdes]
+   [metabase.public-settings :as public-settings]
+   [metabase.util :as u]
+   [metabase.util.date-2 :as u.date]
+   [metabase.util.log :as log]
+   [metabase.util.malli.schema :as ms]
+   [ring.core.protocols :as ring.protocols])
+  (:import
+   (java.io File ByteArrayOutputStream)
+   (java.lang AutoCloseable)
+   (org.apache.commons.compress.archivers.tar TarArchiveEntry TarArchiveInputStream TarArchiveOutputStream)
+   (org.apache.commons.compress.compressors.gzip GzipCompressorInputStream GzipCompressorOutputStream GzipParameters)
+   (org.apache.logging.log4j Level)
+   (org.apache.logging.log4j.core.appender AbstractAppender FileAppender)
+   (org.apache.logging.log4j.core.config AbstractConfiguration LoggerConfig)))
+
+(set! *warn-on-reflection* true)
+
+;;; Request callbacks
+
+(def callbacks* "Callbacks to run on request completion" (atom {}))
+
+(defn- register-cb! [req cb]
+  (swap! callbacks* update req conj cb))
+
+(defn- run-callbacks! [req]
+  (run! #(%) (get @callbacks* req))
+  (swap! callbacks* dissoc req))
+
+(defn- ba-copy [f]
+  (let [baos (ByteArrayOutputStream.)]
+    (io/copy f baos)
+    (.toByteArray baos)))
+
+(defn- on-response! [req data]
+  (reify
+    ;; Real HTTP requests and mt/user-real-request go here
+    ring.protocols/StreamableResponseBody
+    (write-body-to-stream [_ response out]
+      (ring.protocols/write-body-to-stream data response out)
+      (run-callbacks! req))
```

Comment:
I believe `run-callbacks!` will keep the connection until it's completed right?
so:
- how about sending it to a future? 
- do we really need to use an atom here? how about we make `serialize&pack` and `unpack&import` returns a list of callback functions, then we use that as an input for `on-response!`? I don't love the idea of `register-cb!` because it can be called anywhere in the call stack; it makes debugging hard, and it's hard to know what could happen.

Reply:
They will, yeah. I like the idea of passing callbacks directly instead of having an atom, this makes it simpler. I'm not sure about the future though, since removal shouldn't take too much time, and waiting in tests... `with-redefs` won't work since it's macro, so how do I not introduce a flake then?

Reply:
You can use a dynamic var to make it run synchronously in test but async in prod

Reply:
Okay, let's do it like that. Removal takes about 80ms per 1k files at my laptop, so I guess for a slower disk it's going to be even more noticeable.